src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs

```diff
@@ -45,26 +57,44 @@ public void OnSolutionAdded(SolutionInfo solutionInfo)
                 RegisterPrimarySolutionAsync().Wait();
             }
 
+            private async Task<(Connection, bool dispose)> GetConnectionAsync()
+            {
+                this.AssertIsForeground();
+                if (_preferredConnection != null)
+                {
+                    return (_preferredConnection, dispose: false);
+                }
+
+                var connection = await _client.TryCreateConnectionAsync(WellKnownRemoteHostServices.RemoteHostService, CancellationToken.None).ConfigureAwait(false);
+                return (connection, dispose: true);
+            }
+
             private async Task RegisterPrimarySolutionAsync()
             {
                 _currentSolutionId = _workspace.CurrentSolution.Id;
                 var solutionId = _currentSolutionId;
 
-                using (var connection = await _client.TryCreateConnectionAsync(WellKnownRemoteHostServices.RemoteHostService, CancellationToken.None).ConfigureAwait(false))
+                var (connection, dispose) = await GetConnectionAsync().ConfigureAwait(false);
+                try
                 {
                     if (connection == null)
                     {
                         // failed to create connection. remote host might not responding or gone. 
                         return;
                     }
 
-                    await connection.InvokeAsync(
-                        nameof(IRemoteHostService.RegisterPrimarySolutionId), new object[] { solutionId }, CancellationToken.None).ConfigureAwait(false);
+                    var storageLocation = _workspace.DeferredState?.ProjectTracker.GetWorkingFolderPath(_workspace.CurrentSolution);
 
                     await connection.InvokeAsync(
-                        nameof(IRemoteHostService.UpdateSolutionIdStorageLocation),
-                        new object[] { solutionId, _workspace.DeferredState?.ProjectTracker.GetWorkingFolderPath(_workspace.CurrentSolution) },
-                        CancellationToken.None).ConfigureAwait(false);
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId), 
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).ConfigureAwait(false);
+                }
+                finally
+                {
+                    if (dispose)
+                    {
+                        connection?.Dispose();
```

Comment:
I'm not sure that's true.  The old code was written as if ```_client.TryCreateConnectionAsync``` could return null.  So i'm ensuring that if that does happen (either for creating the initial connect, or any subsequent connections) that we're resilient to it.